specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts

```diff
@@ -0,0 +1,440 @@
+  ): void {
+    this.addPromise(
+      this.invokeRule('onRemoved', undefined, [resource, collection])
+    );
+  }
+
+  public setUpManager(): void {
+    this.addPromise(this.invokeRule('customInit', undefined, [this.resource]));
+    if (isTreeResource(this.resource as SpecifyResource<AnySchema>))
+      initializeTreeRecord(this.resource as SpecifyResource<AnyTree>);
+
+    this.resource.on('change', this.changed, this);
+    this.resource.on('remove', this.removed, this);
+  }
+
+  public checkField(fieldName: string) {
+    fieldName = fieldName.toLocaleLowerCase();
+    const thisCheck: Promise<BusinessRuleResult> = flippedPromise();
+    this.addPromise(thisCheck);
+
+    this.fieldChangePromises[fieldName] !== undefined &&
+      this.fieldChangePromises[fieldName].resolve('superseded');
+    this.fieldChangePromises[fieldName] = thisCheck;
+
+    var checks = [
+      this.invokeRule('fieldChecks', fieldName, [this.resource]),
+      this.checkUnique(fieldName),
+    ];
+
+    if (isTreeResource(this.resource as SpecifyResource<AnySchema>))
+      checks.push(
+        treeBusinessRules(this.resource as SpecifyResource<AnyTree>, fieldName)
+      );
+
+    const _this = this;
+
+    Promise.all(checks)
+      .then((results) => {
+        return (
+          thisCheck === _this.fieldChangePromises[fieldName] &&
+          _this.processCheckFieldResults(fieldName, results)
+        );
+      })
+      .then(() => {
+        resolveFlippedPromise(thisCheck, 'finished');
+      });
+  }
+
+  private processCheckFieldResults(
+    fieldName: string,
+    results: RA<BusinessRuleResult | undefined>
+  ): Promise<(void | null)[]> {
+    return Promise.all(
+      results.map((result) => {
+        if (!result) return null;
+        if (result.key === undefined) {
+          if (result.valid)
+            return typeof result.action === 'function' ? result.action() : null;
+          return null;
+        }
+        if (result.valid === false) {
+          this.resource.saveBlockers!.add(result.key, fieldName, result.reason);
+        } else {
+          this.resource.saveBlockers!.remove(result.key);
+          return typeof result.action === 'function' ? result.action() : null;
+        }
+      })
+    );
+  }
+
+  private async checkUnique(fieldName: string): Promise<BusinessRuleResult> {
+    const _this = this;
+    var toOneFields: RA<any> =
+      (this.rules?.uniqueIn && this.rules.uniqueIn[fieldName]) ?? [];
+
+    const results = toOneFields.map((uniqueRule) => {
+      var field = uniqueRule;
+      var fieldNames: string[] | null = [];
+      if (uniqueRule === null) {
+        fieldNames = null;
+      } else if (typeof uniqueRule != 'string') {
+        fieldNames = fieldNames.concat(uniqueRule.otherfields);
+        field = uniqueRule.field;
+      } else fieldNames = [fieldName];
+      return _this.uniqueIn(field, fieldNames);
+    });
+
+    Promise.all(results).then((results) => {
+      _.chain(results)
```

Comment:
Not sure if this is in scope of your refactoring, but here are few more things you could do:
 - Get rid of dependency on underscore.js, jquery.js and other legacy libraries
 - `const _this = this` is likely not needed anymore (it was needed before es6 when annonomous functions didn't exist)
 - If possible, try to factor out common logic into pure helper functions, that are easy to write tests for and are then called by methods in this class